### ISSUE TITLE
Remove twisted.internet.qtreactor

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2130]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2130|
|Type|task|
|Created|2006-09-26 14:21:29Z|

See <http://twistedmatrix.com/pipermail/twisted-python/2006-September/014077.html>.  Summarized version is that PyQt-using applications are GPL, which is incompatible with Twisted's MIT license.

<details><summary>Searchable metadata</summary>

```
trac-id__2130 2130
type__task task
reporter__exarkun exarkun
priority__highest highest
milestone__Core_2_5 Core-2.5
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1159280489000000 1159280489000000
changetime__1159640439000000 1159640439000000
version__None None
owner__ 
cc__edsuom
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun removed owner|
|-|-|

Ready for review in remove-qtreactor-2130

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_2"></a>|@jyknight commented|
|-|-|

Wait, what? 
It's acceptable to have optional parts of a MIT licensed application use GPL code. If you end up depending on said GPL'd parts, it would make your entire program covered by the GPL. And that's okay, as far as Twisted goes, since the MIT license *is* compatible with the GPL. Therefore I disagree that this branch is necessary.

**Comment by twisted-trac** [comment]
|<img alt="edsuom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|edsuom commented|
|-|-|

This comment duplicated in ticket:1770

It's not entirely certain in my mind that a Twisted, MIT-licensed reactor will play nice, license-wise with either a GPL or commercial license to Qt and PyQt.  The best solution might be for [Riverbank Computing](http://www.riverbankcomputing.co.uk/pyqt/) to supply a Twisted reactor module suitable for either GPL or commercial licensing on its own, just like it provides the Python bindings for Qt on that basis.

Meanwhile, however, it might be best to write up a fairly detailed bit of documentation describing how a person can integrate the reactor into Qt's event loop and leave the actual code writing up to the developer.  That would be a real pain, of course. Hopefully Trolltech and Riverbank will see that it's in their best interests to allow people to purchase their commercial-licensed libraries in order to develop commercial PyQt applications with Twisted, without jumping through such a difficult hoop.

**_Disclaimer**_: _Nothing in this comment is to be construed as legal advice. Although I probably think about things legal more often than the average person, being a registered patent agent, I am **not** a lawyer qualified to practice copyright law or anything else outside the representation of patent applicants before the US patent office, and nobody related to this topic is my client in even that capacity._

**Comment by twisted-trac** [comment]
|<img alt="edsuom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|edsuom commented|
|-|-|

Replying to [jknight](#note_2):
> Wait, what? 
> It's acceptable to have optional parts of a MIT licensed application use GPL code. If you end up depending on said GPL'd parts, it would make your entire program covered by the GPL. And that's okay, as far as Twisted goes, since the MIT license *is* compatible with the GPL. Therefore I disagree that this branch is necessary.

The issue as I see it (see disclaimer in previous post) is not PyQt depending on a twisted module, but the reverse. Qtreactor has PyQt dependencies and could be argued to be a derivative work of it, and thus GPL'd:

  Combining two modules means connecting them together so that they form a single
  larger program. If either part is covered by the GPL, ''the whole combination
  must also be released under the GPL''--if you can't, or won't, do that, you may
  not combine them.
  
  What constitutes combining two parts into one program? This is a legal
  question, which ultimately judges will decide. We believe that a proper
  criterion depends both on the mechanism of communication (exec, pipes, rpc,
  function calls within a shared address space, etc.) and the semantics of the
  communication (what kinds of information are interchanged).
  
  If the modules are included in the same executable file, they are definitely
  combined in one program. ''If modules are designed to run linked together in a
  shared address space, that almost surely means combining them into one program.''

From http://www.gnu.org/licenses/gpl-faq.html#TOCMereAggregation

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Replying to [jknight](#note_2):
> Wait, what? 
> It's acceptable to have optional parts of a MIT licensed application use GPL code. If you end up depending on said GPL'd parts, it would make your entire program covered by the GPL. And that's okay, as far as Twisted goes, since the MIT license *is* compatible with the GPL. Therefore I disagree that this branch is necessary.

I worded that poorly.  MIT and GPL are compatible here.  However, GPL modules are not compatible with Twisted's policy, which is that everything is MIT.  Having exceptions to the MIT policy confuses the licensing issue.  It's stupid, yes.  I don't think there's much we can to, though. I also think it would be perfectly fine to continue to distribute the Qt reactor ourselves, just not directly within Twisted.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

The only actual, legal advice I've gotten from a lawyer on the subject of including code from other projects with different licenses and the notion of derivative works with respect to software can be summarized thusly:

 1. Always consult a lawyer and pay them generous fees to explain each particular situation.
 2. If for some reason you can't do the first option, ask for an interpretation of the license from the licenser (the copyright holder) and get that on record in case they ever change their minds.  If it's not clear in any way that what you want to do is allowed, remove the potentially offending code unless you want to be in a legally uncertain area.

Summarized more briefly:

 "When in doubt, take it out."

This seems to jive with ed's interpretation, and while I wouldn't necessarily trust his advice as legally sound (in fact I am required not to :-)) I do trust that he thinks about these issues more than anyone else on the Twisted team.

In this particular case, I'm going to double-check one last thing with Riverbank before reviewing and committing this change.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_7"></a>|@jml set owner to @glyph|
|-|-|

I've reviewed the change. Feel free to merge without further review, pending glyph's answer.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

Thanks for the review, jml.  I've checked with Riverbank and their opinion is that the GPL definitely covers qtreactor, so it should be moved to a separate package.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

Reassigned for merging.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `closed`|
|-|-|

(In [18345]) Merge remove-qtreactor-2130

Author: exarkun
Reviewer: jml
Fixes [#2130](https://github.com/twisted/twisted/issues/2130)
Refs [#2137](https://github.com/twisted/twisted/issues/2137)

This removes support for Qt event loop integration from Twisted, due to
licensing concerns.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
